### ISSUE TITLE
Require #private and #protected to take an argument or not

### ISSUE THREAD

**Issue by adsteel** [rubocop-hq/rubocop#2687]
Currently there is no way to configure Rubocop to raise an error for the following code:

``` ruby
class First
  private def allowed?
    ...
  end
end

class Second
  private

  def failed?
    ...
  end
end
```

It would be nice to enforce a consistent way of declaring private and protected methods across the app.

**Comment by bbatsov** [comment]
And `public`. :-)

**Comment by envygeeks** [comment]
What you would do in the situation where you see somebody use `private_class_method def` and then `private\ndef` instead of `private def`?  `{private,public}_class_method` doesn't get inherited (and I don't remember if there is a way to get it to inherit)

**Comment by rrosenblum** [comment]
`private` and `private_class_method` do not work exactly the same. When creating a class that only has static methods, in order to safe guard someone from trying to instantiate a useless instance of the class, I will use `private_class_method :new`. Trying to set `private :new` will result in a `NameError` unless `new` has been defined by us.

```ruby
# does not work
class Foo
  def self.bar
    puts 'bar'
  end
  private :new
end

# => NameError: undefined method `new' for class `Foo'
# => Did you mean?  new
```

```ruby
# works
class Foo
  def self.bar
    puts 'bar'
  end
  private_class_method :new
end
```

Also, keep in mind that `def` does not return the method name until Ruby 2.0. While the cop could still apply, I find that passing arguments to `private, public, protected` to be rather ugly in Ruby 1.9.3, and that the no argument version should be preferred.

```ruby
# Ruby 1.9.3
def foo
end
private :foo
```

Generally speaking, I personally prefer the no argument version. It works well unless you have a very large class. When that happens, the modifiers tend to get lost and it isn't as easy to identify what is `private` and `protected`. These large classes tend to be violating other core concepts and should be refactored. In the reality, it isn't always easy to do so due to complexity or time constraints. It might be a good idea to have `consistent` setting (similar to `TrailingComma`) which makes sure that access modifiers are used consistently in a file.

**Comment by envygeeks** [comment]
> private and private_class_method do not work exactly the same. When creating a class that only has static methods, in order to safe guard someone from trying to instantiate a useless instance of the class, I will use private_class_method :new. Trying to set private :new will result in a NameError unless new has been defined by us.

That's because `private_class_method` operates on the singleton, and `private` operates on the instance.  All you did was highlight why `private_class_method` exits, you didn't highlight any non-user-driven mistake difference.  My comment was about:

```ruby
[1] pry(main)> class MyClass
[1] pry(main)*   private_class_method def self.hello_world
[1] pry(main)*     "hello world"
[1] pry(main)*   end  
[1] pry(main)* end
=> MyClass

[2] pry(main)> MyClass.hello_world
NoMethodError: private method `hello_world' called for MyClass:Class
from (pry):6:in `__pry__'
```

**Comment by rrosenblum** [comment]
Thank you for the clarification. I didn't fully comprehend your initial comment. The use case of `private_class_method def` combined with `private` without args brings up a good stylistic approach. 

```ruby
class MyClass
  private_class_method def self.hello_word
    'hello world'
  end
end
```
could also be written as
```ruby
class MyClass
  class << self
    private

    def hello_word
      'hello world'
    end
  end
end
```
The cop could suggest or allow `private_class_method def` when `private def` is used, and suggest 
```ruby
class << self
  private def ...
end
```
when 
```ruby
private

def ...
```
is used.

**Comment by ptarjan** [comment]
If this cop gets implemented, we'd like a mode where `private` was forced to appear before the method instead of just somewhere randomly in the file.

```
# bad
class Foo
  private

  def bar
  end

  def baz
  end
end
```

```
# good
class Foo
  private def bar
  end

  private def baz
  end
end
```

Autocorrecting all our existing code is pretty important for this cop.

**Comment by Darhazer** [comment]
`Style/AccessModifierDeclarations` handles this
